src/converters/ssm-dynamic.ts

```diff
@@ -0,0 +1,160 @@
+ */
+export function resolveSSMDynamicPlaintextReference(
+    parent: pulumi.Resource,
+    value: string,
+): pulumi.Output<string | string[]> {
+    // This shouldn't happen because we currently only call this where we know we have a string
+    // but adding this for completeness
+    if (containsEventuals(value)) {
+        throw new Error('SSM dynamic references cannot contain unresolved values');
+    }
+
+    const parts = parseSSMDynamicPlaintextReference(value);
+    return aws.ssm
+        .getParameterOutput(
+            {
+                name: parts.parameterName,
+                // we don't want to return a decrypted SecureString value
+                // SecureString types are handled elsewhere
+                withDecryption: false,
+            },
+            { parent },
+        )
+        .apply((v) => {
+            switch (v.type) {
+                // CDK/CloudFormation will return a string for both String and StringList types
+                case 'String':
+                case 'StringList':
+                    return v.value;
+                default:
+                    throw new Error(`Unsupported SSM parameter type: ${v.type}`);
+            }
+        });
+}
+
+/**
+ * Resolves an SSM SecureString dynamic reference
+ *
+ * @param parent - The parent resource for the SSM parameter function
+ * @param value - The value which contains the SSM SecureString dynamic reference
+ * @returns The parameter value as a pulumi secret output
+ */
+export function resolveSSMDynamicSecureStringReference(parent: pulumi.Resource, value: string): pulumi.Output<string> {
+    // This shouldn't happen because we currently only call this where we know we have a string
+    // but adding this for completeness
+    if (containsEventuals(value)) {
+        throw new Error('SSM dynamic references cannot contain unresolved values');
+    }
+
+    const parts = parseSSMDynamicSecureStringReference(value);
+    return aws.ssm
+        .getParameterOutput(
+            {
+                name: parts.parameterName,
+                withDecryption: true,
+            },
+            { parent },
+        )
+        .apply((v) => {
+            switch (v.type) {
+                case 'SecureString':
+                    return pulumi.secret(v.value);
+                default:
+                    throw new Error(`Unsupported SSM parameter type: ${v.type}`);
+            }
+        });
+}
+
+/**
+ * Used to process a value that may contain a ssm dynamic reference
+ *
+ * The value may be a pulumi output (typically if the value contains resource references) or a string.
+ *
+ * @param parent - The parent resource
+ * @param value - A fully resolved value that may contain a ssm dynamic reference
+ * @returns A secret output if the value is a ssm dynamic reference, otherwise the original value
+ */
+export function processSSMReferenceValue(parent: pulumi.Resource, value: any): any {
+    let returnValue = value;
```

Comment:
Possibly better to just use `lift`? It also will take care of promises, not just Output values, which do arise.

```typescript
const f = (value: any) => {
    // safely assume value is not an output here
    if (typeof value === 'string) {
       // try transforming refs
    } else {
       return value;
    }
};
return lift(f, value);
```

Reply:
Using lift requires some additional knowledge/handling of the types going into lift (e.g arrays need to be handled as arrays). I took the necessary pieces from lift to make this better.